File: frontend/src/services/realmService.ts
import { request } from '@umijs/max';

export interface RealmDto {
  id?: string;
  realm: string;
  displayName?: string;
  enabled: boolean;
  loginTheme?: string;
  accountTheme?: string;
  adminTheme?: string;
  emailTheme?: string;
}

export async function getAllRealms(): Promise<RealmDto[]> {
  return request('/api/realms', {
    method: 'GET',
  });
}

export async function getRealm(realmName: string): Promise<RealmDto> {
  return request(`/api/realms/${realmName}`, {
    method: 'GET',
  });
}

export async function createRealm(realm: RealmDto): Promise<RealmDto> {
  return request('/api/realms', {
    method: 'POST',
    data: realm,
  });
}

File: backend/src/main/kotlin/com/nova/iam/service/RealmService.kt
package com.nova.iam.service

import com.nova.iam.dto.RealmDto
import org.keycloak.admin.client.Keycloak
import org.keycloak.representations.idm.RealmRepresentation
import org.springframework.stereotype.Service

@Service
class RealmService(private val keycloak: Keycloak) {

    fun getAllRealms(): List<RealmDto> {
        return keycloak.realms().findAll().map { realm ->
            RealmDto(
                id = realm.id,
                realm = realm.realm,
                displayName = realm.displayName,
                enabled = realm.isEnabled ?: true,
                loginTheme = realm.loginTheme,
                accountTheme = realm.accountTheme,
                adminTheme = realm.adminTheme,
                emailTheme = realm.emailTheme
            )
        }
    }

    fun getRealm(realmName: String): RealmDto? {
        return try {
            val realm = keycloak.realm(realmName).toRepresentation()
            RealmDto(
                id = realm.id,
                realm = realm.realm,
                displayName = realm.displayName,
                enabled = realm.isEnabled ?: true,
                loginTheme = realm.loginTheme,
                accountTheme = realm.accountTheme,
                adminTheme = realm.adminTheme,
                emailTheme = realm.emailTheme
            )
        } catch (e: Exception) {
            null
        }
    }

    fun createRealm(realmDto: RealmDto): RealmDto {
        val realmRepresentation = RealmRepresentation().apply {
            realm = realmDto.realm
            displayName = realmDto.displayName
            isEnabled = realmDto.enabled
            loginTheme = realmDto.loginTheme
            accountTheme = realmDto.accountTheme
            adminTheme = realmDto.adminTheme
            emailTheme = realmDto.emailTheme
        }
        
        keycloak.realms().create(realmRepresentation)
        
        return getRealm(realmDto.realm) ?: throw RuntimeException("Failed to create realm")
    }
}